From e4bda5413c278c57cc21647a768da297841719ed Mon Sep 17 00:00:00 2001
From: Amadeusz Skrzypczak <amadeusz.skrzypczak@redembedded.com>
Date: Thu, 1 Apr 2021 09:25:51 +0200
Subject: [PATCH] Fix timeout in WebProcess during browser exit

Allow the sync message ShouldTerminate to be answered by the UI process
even if the UI process is waiting for a sync message reply.
Cherry-pick from upstream WPEWebkit
https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/9dc56ed517c5ea08fb007cab60eb044ddbc6a709

---
 Source/WebKit/Platform/IPC/Connection.h | 4 ++++
 Source/WebKit/WebProcess/WebProcess.cpp | 2 +-
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/Source/WebKit/Platform/IPC/Connection.h b/Source/WebKit/Platform/IPC/Connection.h
index 039253b9..79b33b9f 100644
--- a/Source/WebKit/Platform/IPC/Connection.h
+++ b/Source/WebKit/Platform/IPC/Connection.h
@@ -67,6 +67,7 @@ enum class SendSyncOption {
     // Use this to inform that this sync call will suspend this process until the user responds with input.
     InformPlatformProcessWillSuspend = 1 << 0,
     UseFullySynchronousModeForTesting = 1 << 1,
+    DispatchMessageEvenWhenWaitingForSyncReply = 1 << 2,
 };
 
 enum class WaitForOption {
@@ -441,6 +442,9 @@ template<typename T> bool Connection::sendSync(T&& message, typename T::Reply&&
         m_fullySynchronousModeIsAllowedForTesting = true;
     }
 
+    if (sendSyncOptions.contains(SendSyncOption::DispatchMessageEvenWhenWaitingForSyncReply))
+        encoder->setShouldDispatchMessageWhenWaitingForSyncReply(true);
+
     // Encode the rest of the input arguments.
     encoder->encode(message.arguments());
 
diff --git a/Source/WebKit/WebProcess/WebProcess.cpp b/Source/WebKit/WebProcess/WebProcess.cpp
index 15de96b9..4bf08fca 100644
--- a/Source/WebKit/WebProcess/WebProcess.cpp
+++ b/Source/WebKit/WebProcess/WebProcess.cpp
@@ -608,7 +608,7 @@ bool WebProcess::shouldTerminate()
 
     // FIXME: the ShouldTerminate message should also send termination parameters, such as any session cookies that need to be preserved.
     bool shouldTerminate = false;
-    if (parentProcessConnection()->sendSync(Messages::WebProcessProxy::ShouldTerminate(), Messages::WebProcessProxy::ShouldTerminate::Reply(shouldTerminate), 0)
+    if (parentProcessConnection()->sendSync(Messages::WebProcessProxy::ShouldTerminate(), Messages::WebProcessProxy::ShouldTerminate::Reply(shouldTerminate), 0, Seconds::infinity(), IPC::SendSyncOption::DispatchMessageEvenWhenWaitingForSyncReply)
         && !shouldTerminate)
         return false;
 
-- 
2.17.1

